### DISCUSSION TITLE
location_notes created from a TSML JSON feed do not update when the feed updates?

### DISCUSSION THREAD

**Discussion by brotskydotcom2** [code4recovery/12-step-meeting-list#1235] [Q&A]
Our meeting list includes [a tsml json feed](https://district26-production-cfb60bb59f05.herokuapp.com/api/tsml) for an Al-Anon district.  That feed defines meetings at several locations which are used only by meetings in that feed.

The current behavior of the `wp_ajax_tsml_import` action, on [lines 433-448 of ajax.php](https://github.com/code4recovery/12-step-meeting-list/blob/7fb05b9b2b7b4547e0bf6b40d84696deda20e02a/includes/ajax.php#L433C1-L449C1), is to prefer cached location data (based on address) over the data in the feed about a meeting.  This means that if the feed has updated the `location_notes` field, the import will not update this value.

This feels like a bug to me, especially in the situation where there are no meetings at that location other than in this feed.  And there seems to be no workaround, other than deleting all these locations from the database before refreshing the feed.

Given that locations are posts of type `tsml_location`, it seems to me that a more conservative caching strategy might be to use the cache only when the location last-modified date is greater than or equal to the meeting update date.  A simpler but more brute force fix would just be always to update the location_notes from the feed.

I am happy to contribute either of these fixes, but first wanted to ask the community whether I'm missing something obvious (e.g., that there is a way to force the update when defining the feed) or whether there's another reason why the code works this way.

**Comment by kiyote33** [comment]
Hi @clickonetwo, I think there is an update in the pipeline for this issue. In the meantime, you could force a refresh of the cache file by editing any meeting and simply update the record without changing anything. That will force a refresh of the cache file. Let me know if that solves your problem for now.

**Comment by joshreisner** [comment]
yeah, those lines do look problematic to me, but it also goes deeper than that. TSML makes locations unique by address, so two locations (imported or not) that share an address will also by necessity share fields like `location`, `region`, and `location_notes`

the feed you are importing from does not obey that logic, so even when we fix that code, by necessity you will still be losing some location notes when importing from that feed. 

for example, `Morning Light Women's Meeting` is in `Berkeley, CA` and has location notes of `Enter the church from the (free) parking lot.  We meet downstairs (same level as the door).`, however, there are several other meetings in Berkeley CA that have no location notes, so I think they are colliding. 

(note that TSML considers it an address even from an approximate location like `Berkeley, CA`)

the fix for this is to ask the website you're importing from to move the location notes to the regular (meeting) `notes` field. these can vary by meeting, while `location_notes` apply to every meeting at that location, and `group_notes` apply to every meeting with that group name

**Reply by brotskydotcom2** [reply]
Hi @kiyote33 and @joshreisner thanks much for your responses!

Yes, I'm aware of the issue about locations being standardized by geolocated address, and I'm working with the District to make sure that their location notes are standardized that way. (They had thought they were standardized by location name, not by address.) I am also working with them to use meeting notes where necessary.

**Comment by brotskydotcom2** [comment]
Thanks for the advice! It still feels to me like there a bug in this code, because I believe that any mechanism capable of creating a location should be capable of updating that location.  This bug is compounded by the fact that location notes can't be updated directly in the admin UI; they have to be updated in the context of a specific meeting at that location.

I would like to suggest two approaches to addressing this issue, and see whether either of them appeals to you:

1. (Simpler) If the `location_notes` in the feed differ from the existing (database) location notes, update the db.
2. (More Flexible) Offer an option with feed definition that deletes locations in the feed so they are all created fresh by the feed. (This is similar to the options offered with CSV import.)

If you like the simpler of these, I am happy to file a bug and create a PR for you to review.   If you prefer the more complicated one, which would involve UI changes and touching a lot more places in the code, I am willing to give it a shot but I am sure I would need some mentoring.

**Reply by joshreisner** [reply]
thanks @clickonetwo i think we need to talk about it as a group - we'd love to have you join us sometime! (our next meeting is saturday 10am Pacific Time)

i'm not sure it's actually not working properly. if website B is imported into website A, and then B updates location notes, then they will be updated on website A upon the next import *if A doesn't also list meetings at that location*

this happens because (similar to your option 2 above) before the import runs, A deletes all the meetings imported from B *and all the newly-"orphaned" (bad term) locations* so they will be reimported fresh. i suspect we'll do a live demo of this on saturday, but i just ran it myself so i know it definitely works. 

and also as @kiyote33 mentioned above, there is an effort underway to refactor the remote JSON import algorithm, so I wouldn't recommend starting any work on this at this time

**Reply by brotskydotcom2** [reply]
Thanks @joshreisner I was also able to confirm that, if a location only has meetings from one feed, and that feed is refreshed, that location is deleted and then recreated from the feed.  I think the problem we were seeing with the `location_notes` was due to earlier attempts to import those district meetings via CSV.  Those meetings were deleted but the trash was not emptied, so those meeting locations were never orphaned and deleted when the feed was refreshed.

So that just leaves the case where the location belongs to multiple feeds (or to a feed and an internally created meeting).  In that case I can see that not updating the location notes is justified, but I believe it should at least be flagged as a non-fatal import error so the failure to update is not invisible.  Maybe that can be picked up in the refactor that's going on.

Thanks much for inviting me for Saturday. I'm sorry but I have a conflict so I can't make it.
